### ISSUE TITLE
A way to “unload” plugins

### ISSUE THREAD

**Issue by uranusjr** [LeaVerou/prism#459]
I have a project using Prism to perform server-side rendering (similar to node). Simple rendering works well, but I’m facing some problem trying to use plugins. A plugin can be loaded simply by running the appropriate JavaScript file, but there seems to be no way to undo the process. Say I want to render two code blocks, first one with line numbers, and second one without—there is no way to do this unless I construct a new Prism instance for each one of them. Is it possible to add some kind of mechanism possible?

**Comment by Golmote** [comment]
I have several ideas for this, but each has its own issues.

We could ask the plugins to input a third parameter when they add hooks:
```javascript
Prism.hooks.add('after-highlight', function (env) { /* ... */ }, 'line-numbers');
```
so that we can do something like `Prism.hooks.removePlugin('line-numbers');` to remove all hooks added by this plugin.

Unfortunately, some plugins also modify the language definitions to add/change some tokens. So another mechanism may be needed.
We could ask the plugins to "register" themselves to Prism, by providing a `load` and an `unload` method. This may be combined with the first idea.
```javascript
Prism.plugins.add('autolinker', function() {
    // Code to init plugin here
    for (var language in Prism.languages) {
        /* ... */
    }
    Prism.hooks.add('wrap', function (env) { /* ... */ }, 'autolinker');
}, function () {
    // Code to unload plugin here
    for (var language in Prism.languages) {
        // Undo what was done in init
    }
    Prism.hooks.removePlugin('autolinker');
});
```
Hence giving us the ability to provide something like `Prism.disablePlugin('autolinker');` (and even `Prism.enablePlugin('autolinker');` to enable it again).

But another issue will be related to the CSS files added by the plugins. I don't see a way to easily apply the styling only on one code block and not the other, especially if the styling applies to tokens that were not added by the plugin...

**Comment by vkbansal** [comment]
I also wanted to add this functionality to my port and I swear I too had similar ideas in my mind. In fact I've Implemented couple of plugins using an approach similar to your first approach.
https://github.com/vkbansal/prism.php/blob/master/src/Components/Plugin/LineNumbers.php

But while I was looking at `show-invisibles` plugin, I too was set back by seeing that it modified languages. A way around it is that plugins should modify `env.grammar`,and not the original language definitions. But again the first approach would allow each plugin to add one and only one hook. What if a plugin wants to add multiple hooks?

I like your second approach but will need to check it out.

As for the styling is considered, you are correct, adding a class will tackle the problem. I've used the very same technique in the port.

**Comment by LeaVerou** [comment]
Asking plugin authors to register load and unload events will discourage many from plugin development I’m afraid :(

**Comment by vkbansal** [comment]
@Golmote I tested your second proposal and It's apt (at least for my project). Though I had to change how hooks were registered so as to keep track of which plugin is adding which hook and how to remove it.

Code:
https://github.com/vkbansal/prism.php/blob/master/src/Components/Plugin/LineNumbers.php

**Comment by Golmote** [comment]
@vkbansal Great to see it's apt. Unfortunately, if @LeaVerou considers this will be too much work for the plugins developers, we'll have to think differently for Prism JS.

I guess it would be quite good already if we only added the third parameter on `hooks.add()` and the CSS classes to restrict the styling...

**Comment by vkbansal** [comment]
@Golmote How about namespaced listeners similar to jQuery?

For example `before-highlight.show-linenumbers` should be

```
hooks = {
    'before-highlight': {
        'show-linenumbers': callback()
    }
}
```

removing hooks should be straight-forward then.

**Comment by Golmote** [comment]
Yes, why not. I'm not sure I see the real difference with something like
```javascript
Prism.hooks.add('after-highlight', function (env) { /* ... */ }, 'line-numbers');
```
but I don't mind.

**Comment by vkbansal** [comment]
Yes your are right. It popped in my head and just wanted to share.

**Comment by pmlk** [comment]
Chiming in ten years later as I'm looking for a similar feature. :) 

Since @uranusjr asked specifically about line numbers, it seems that the [specific line numbers plugin](https://prismjs.com/plugins/line-numbers/) does support enabling and disabling it, simply by adding `class="line-numbers"` or `class="no-line-numbers"` to any parent. This gives quite fine grained control as you can enable it globally with `<body class="line-numbers">` and then disable it for a specific block of code or a section `<div class="no-line-numbers">`.

If I'm understanding this correctly, it seems to be similar to this approach...

> EDIT: About the styling issue, we could in fact add a CSS class on the code block for each plugin that were enabled at the moment it was highlighted. Then, the CSS rules used in the plugin's stylesheet could be applied only to the blocks concerned.

... only that instead of the plugin adding a CSS class after highlighting, the plugin adds its styling only to those code blocks with a given CSS class somewhere in the hierarchy.

It would be awesome if this would hold true for all plugins. Specifically, I'd like to enable the [show-invisibles plugin](https://prismjs.com/plugins/show-invisibles/) only in specific code blocks, e.g. for Makefile, since in Makefiles the distinction between tabs and spaces is quite important. Anybody have an idea how to best accomplish this, now, ten years later? :)

**Comment by pmlk** [comment]
For my specific case I hacked this together:

```js
// selectably-show-invisibles.js
document.addEventListener("DOMContentLoaded", () => {
  document.querySelectorAll('pre > code').forEach(code => {
    const isInsideShowInvisibles = code.closest('.show-invisibles');

    if (!isInsideShowInvisibles) {
      code.querySelectorAll('.token.space, .token.tab, .token.cr, .token.lf').forEach(el => {
        el.replaceWith(document.createTextNode(el.textContent));
      });
    }
  });
});
```

Then

```html
  ...
  <script scr="prism.js"></script>
  <script scr="selectably-show-invisibles.js"></script>
</body>
```

It just removes the stuff the plugin added UNLESS the code block is inside any element with `class="show-invisibles"`. Not great but it works for me for now.

**Comment by LeaVerou** [comment]
I think we need to disentangle two things:
1. By “unloading a plugin”, I’m imagining that all side effects of the plugin would be rolled back. In v2 one can create multiple Prism instances with different plugins, so a way to facilitate a new instance based on an existing one could be a possible direction. Or the unload method @golmote suggested, as long as it’s optional (some side effects cannot be rolled back after all).  
2. What @pmlk is describing in the last post is not plugin unloading, it’s scoping/granularity and should absolutely be possible with any plugin where it makes sense. I would consider that a plugin feature, not a Prism feature, though Prism could provide helpers to make the pattern easier.
